storage/watchable_store.go

```diff
@@ -241,57 +242,111 @@ func (s *watchableStore) syncWatchingsLoop() {
 	}
 }
 
-// syncWatchings syncs the watchings in the unsyncd map.
 func (s *watchableStore) syncWatchings() {
+	if len(s.unsynced) == 0 {
+		return
+	}
+
+	// in order to find key-value pairs from unsynced watchings, we need to
+	// find min/maximum revision index, and these revisions can be used to
+	// query the backend store of key-value pairs
+	minRev, maxRev := int64(math.MaxInt64), int64(math.MinInt64)
+
+	// get the main revision of the last compaction.
+	// if unsynced watching's channel is not full and its
+	// watching revision is beyond current revision, move
+	// it to synced and watch its future events.
 	_, curRev, _ := s.store.Range(nil, nil, 0, 0)
+
+	keyToUnsynced := make(map[string]map[*watching]struct{})
+
 	for w := range s.unsynced {
-		var end []byte
-		if w.prefix {
-			end = make([]byte, len(w.key))
-			copy(end, w.key)
-			end[len(w.key)-1]++
-		}
+
 		limit := cap(w.ch) - len(w.ch)
-		// the channel is full, try it in the next round
 		if limit == 0 {
+			// channel is full, so try at next round
 			continue
 		}
-		revbs, kvs, nextRev, err := s.store.RangeHistory(w.key, end, int64(limit), w.cur)
-		if err != nil {
-			// TODO: send error event to watching
+
+		k := string(w.key)
+
+		if w.cur > curRev {
+			if err := unsafeAddWatching(&s.synced, k, w); err != nil {
+				log.Panicf("error unsafeAddWatching (%v) for key %s", err, k)
+			}
 			delete(s.unsynced, w)
 			continue
 		}
 
-		// push events to the channel
-		for i, kv := range kvs {
-			var evt storagepb.Event_EventType
-			switch {
-			case isTombstone(revbs[i]):
-				evt = storagepb.DELETE
-			default:
-				evt = storagepb.PUT
-			}
+		if minRev >= w.cur {
+			minRev = w.cur
+		}
+		if maxRev <= w.cur {
+			maxRev = w.cur
+		}
+
+		if _, ok := keyToUnsynced[k]; !ok {
+			keyToUnsynced[k] = make(map[*watching]struct{})
+		}
+		keyToUnsynced[k][w] = struct{}{}
+	}
 
+	// to prevent returning nothing
+	if minRev == maxRev {
+		maxRev = maxRev + curRev
```

Comment:
I was doing this part wrong... It should have added `curRev` to have enough range.